server/bleep/src/webserver/answer.rs

```diff
@@ -71,20 +75,48 @@ pub struct AnswerResponse {
 const SNIPPET_COUNT: usize = 13;
 
 pub async fn handle(
+    params: Query<Params>,
+    Extension(app): Extension<Application>,
+) -> Result<impl IntoResponse, impl IntoResponse> {
+    // create a new analytics event for this query
+    let mut event = QueryEvent::default();
+
+    // populate analytics event
+    let response = _handle(params, Extension(app.clone()), &mut event).await;
+
+    // send event to rudderstack
+    app.track_query(event);
+
+    response
+}
+
+pub async fn _handle(
     Query(params): Query<Params>,
     Extension(app): Extension<Application>,
+    analytics_event: &mut QueryEvent,
 ) -> Result<impl IntoResponse, impl IntoResponse> {
+    let query_id = uuid::Uuid::new_v4();
+
     let semantic = app
         .semantic
         .as_ref()
         .ok_or_else(|| super::error(ErrorKind::Configuration, "Qdrant not configured"))?;
 
+    analytics_event.user_id = params.user_id.clone();
+    analytics_event.query_id = query_id;
+    analytics_event.overlap_strategy = semantic.overlap_strategy();
+
+    analytics_event
+        .stages
+        .push(Stage::new("user query", params.q.clone()));
+
     let query =
         parser::parse_nl(&params.q).map_err(|e| super::error(ErrorKind::User, e.to_string()))?;
     let target = query
         .target()
         .ok_or_else(|| super::error(ErrorKind::User, "missing search target".to_owned()))?;
 
+    let start_semantic_results = Instant::now();
```

Comment:
these duration variables have poorly thought out names, any suggestions?

Reply:
Can you maybe wrap these in a function instead of dumping it all in one namespace? It may be easier to see what's being measured as well, e.g

```
analytics_event.timed("semantic results", || {
... do something
})?
```

or even introduce a `struct` as state that has processing stages and internally mutates.

```
analytics_event.timed("semantic results", context.next_step);
```

Reply:
the first suggest is obviously neater, but i think it requires a greater refactor of that whole module. i have opted to introduce a new timekeeping structure instead, and track the time-elapsed between the start and stop of each stage.

Reply:
This looks a lot cleaner! 👍